Wikipedia/View Controllers/ShareCard/WMFShareOptionsController.m

```diff
@@ -0,0 +1,349 @@
+//  Copyright (c) 2015 Wikimedia Foundation. All rights reserved.
+//
+
+#import "WMFShareOptionsController.h"
+
+#import "Wikipedia-Swift.h"
+#import <Masonry/Masonry.h>
+
+#import "WMFShareFunnel.h"
+
+#import "NSString+Extras.h"
+#import "NSString+WMFHTMLParsing.h"
+
+#import "UIView+WMFSnapshotting.h"
+
+#import "WMFShareCardViewController.h"
+#import "WMFShareOptionsView.h"
+#import "PaddedLabel.h"
+#import "WikipediaAppUtils.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface WMFShareOptionsController ()<UIPopoverControllerDelegate>
+
+@property (strong, nonatomic, readwrite) MWKArticle* article;
+@property (strong, nonatomic, readwrite) WMFShareFunnel* shareFunnel;
+
+@property (nullable, copy, nonatomic) NSString* snippet;
+@property (weak, nonatomic) UIViewController* containerViewController;
+@property (nullable, weak, nonatomic) UIBarButtonItem* originButtonItem;
+@property (nullable, weak, nonatomic) UIView* originView;
+
+@property (nullable, strong, nonatomic) UIView* grayOverlay;
+@property (nullable, strong, nonatomic) WMFShareOptionsView* shareOptions;
+@property (nullable, strong, nonatomic) UIImage* shareImage;
+
+@property (nullable, strong, nonatomic) UIPopoverController* popover;
+
+@end
+
+@implementation WMFShareOptionsController
+
+- (void)cleanup{
+    
+    [[WMFImageController sharedInstance] cancelFetchForURL:[NSURL wmf_optionalURLWithString:self.article.imageURL]];
+    self.containerViewController = nil;
+    self.originButtonItem = nil;
+    self.originView = nil;
+    self.shareImage = nil;
+    self.snippet = nil;
+}
+
+
+- (instancetype)initWithArticle:(MWKArticle*)article
+                    shareFunnel:(WMFShareFunnel*)funnel{
+    
+    NSParameterAssert(article);
+    NSParameterAssert(funnel);
+    NSParameterAssert(article.title.desktopURL.absoluteString);
+
+    self = [super init];
+    
+    if (self) {
+        _article        = article;
+        _shareFunnel = funnel;
+    }
+    return self;
+}
+
+#pragma mark - Public Presentation methods
+
+- (void)presentShareOptionsWithSnippet:(NSString*)snippet inViewController:(UIViewController*)viewController fromBarButtonItem:(nullable UIBarButtonItem*)item{
+    self.snippet = [snippet copy];
+    self.containerViewController = viewController;
+    self.originButtonItem = item;
+    self.originView = nil;
+    [self fetchImageThenShowShareCard];
+}
+
+
+- (void)presentShareOptionsWithSnippet:(NSString*)snippet inViewController:(UIViewController*)viewController fromView:(nullable UIView*)view{
+    self.snippet = [snippet copy];
+    self.containerViewController = viewController;
+    self.originButtonItem = nil;
+    self.originView = view;
+    [self fetchImageThenShowShareCard];
+}
+
+#pragma mark - Asynchornous Fetch and Present
+
+- (void)fetchImageThenShowShareCard {
+    @weakify(self);
+    [[WMFImageController sharedInstance] fetchImageWithURL:[NSURL wmf_optionalURLWithString:self.article.imageURL]].then(^(WMFImageDownload* download){
```

Comment:
two comments about this:

1. Using `fetchImageWithURL:` unfortunately prevents this from showing immediately in the case that the image is already cached in memory
2. One of my gripes w/ how sharing is currently implemented is that we don't do anything until we have _all_ the data, which is exacerbated by 1.

Both of these result in a noticeable lag when the user taps the share button.  So, I'm going to file another ticket to fix this lag by displaying the share options immediately and lazily load the image using the `UIImageView+MWKImage` category.

Unrelated, but I'm also noticing some UX things that I'd like to bring up w/ design, namely:
- Bouncing the share button seems out of place. Nothing else in our UI does this, and I think it will seem out of place or distracting if we're presenting the share options immediately as I outlined above
- Adding a (red) cancel button to the options. All other OS sheets have one, and I expect that user testing would reveal users getting confused and just tapping the screen out of desperation to get out, instead of having a clear exit path.
Adding the cancel button is lower priority, but I think the removal of bouncing is has low enough cost:benefit ratio to do now if we agree it should be done.